### PR TITLE
PWND.

### PR DESCRIPTION
**Pull request by jboxman**
While this mostly works, the CSS is a disaster, as is the available space. Between both TOCs, where does the content go?

### CODE REVIEW

**Comment by jboxman** [comment]
So there are realities to confront with this:

- Mobile wasn't designed for this; Is it possible to hide it using a media query? What about if a user has JavaScript disabled?
- Is it possible to account for the height of the header dynamically? That is, once a user scrolls, how many pixels must pass until the widget pops out from the page? (The `display` is changed to `fixed` from `static`.) Right now, a brute for specific pixel height is used, but that might not be useful in practice.
- Once the element transitions to `display: fixed`, it pops out of the normal documentation flow and no longer respects the container width; this means it fills up the entire remainder of the page on the right side. So it jumps out once the user scrolls far enough to change to `display: fixed`. Can this be mitigated? I have no idea.
- The layout wasn't designed for 3 columns, so there isn't enough space available. How small can the text be for the left TOC? For the new right TOC? For the content column? If we try to expand the available space overall, what happens then? It's set for a fixed 768 width or 1024 width, I think. (i.e. not fluid; and as near as I can tell, transitioning to a fluid layout means a redesign of the entire site layout from scratch.)

In short, the stylesheet that backs the layout supports some of this, but the d.o.c layout might not without revising it.

What are the unintended consequences of this?

**Comment by IronicBadger** [comment]
This is a great start!

In my opinion, mobile support for this would not be expected. Where would the content go!? 👍 

In terms of spacing Gitbook handle this very nicely, although they possible have a bit too much whitespace in their primary left nav menu - https://docs.gitbook.com/.

**Comment by vikram-redhat** [comment]
> What are the unintended consequences of this?

Let's try this in a limited release, for perhaps, the 4.6 docs?

**Comment by jboxman** [comment]
The Gitbook example vastly exceeds my proficiency with this.

If I had time to upload a working demo, there are still some rather serious UX issues with the implementation in this PR such that it appears amateurish.

I can certainly experiment with it, but I don't think it's practical to include anything right now prior to docs freeze.

**Comment by jboxman** [comment]
Our style sheet is an ancient float-based deal. The TOC app web site for example uses modern dynamically calculated CSS properties, and that makes it trivial to keep a consistent look when the TOC transitions from within the normal document flow to `position: fixed` relative to the browser viewport.

I have no idea how to replicate this for our content with the stylesheet we have in hand.

I'd need to reverse engineer exactly how our stylesheet uses floats and percentage widths, and maybe it's possible to achieve a similar result. I'm not sure. When the TOC is set to `fixed`, it flushes left, because of the floats. The traditional fix for this seems to be to set a `margin-left` of identical size, but this somehow needs to account for paddings and margins of the other columns.

Maybe it can't be done, I don't know.

Anything less and the TOC content obviously and glaringly repositions once the viewport scrolls past the header master.

**Comment by jboxman** [comment]
I don't think this is possible.

We're using a % based layout, with pixel fixed padding.

Once the `fixed` position is applied, somehow it is necessary to account for the % and fixed width spacing and padding on a div that's been yanked out of the normal document flow so that it can line up with where it was when it existed within the normal flow, and was thus subject to the padding and width of the layout. As far as I know, with legacy CSS this is not possible.

This would all be trivia with CSS layouts, but that was not widely supported until 3-4 years ago, and we probably can't redo the column layouts to accomplish this. (That said, my knowledge of CSS is mostly limited to the level of _change the color of some text_, so I might be wrong.)

Can always have the TOC just magically appear once the user has scrolled far enough; that might be jarring, though.

**Comment by jboxman** [comment]
Or I could try replacing the fixed pixel with padding with percentage widths, governed by media queries to adjust based on screen with as it already does anyway. If that doesn't work, my assessment about this is probably wrong anyway. But if it works it's probably a small enough change to be viable.

**Comment by jboxman** [comment]
@vikram-redhat @IronicBadger 

Seems to work so far; need to find a way to hide it on mobile/extra-small size, because it isn't really viable.

https://codepen.io/jboxman/full/YzWqYME

**Comment by jboxman** [comment]
Comically it actually _does_ work as is on IE 9. Works on latest Firefox. Works on recent Chromium.

**Comment by jboxman** [comment]
And works with mobile sizes now, but simply hiding the column. Mobile is a single column view anyway, so no additional magic is necessary besides a single media query.

**Comment by jboxman** [comment]
However, the container invisibly blots out the content, so either the content elements are not selectable or the TOC links are not selectable. It might be there is no fix for that, and that means this is useless.

**Comment by jboxman** [comment]
Using a fixed position relative to the right side of the container seems to solve this, cleanly, but it doesn't seem possible to cleanup the transition. It'll reflow awkwardly because the container size isn't a constant, it changes once the element is fixed.

But at least it's possible to interact with the page contents and the links this way.

**Comment by jboxman** [comment]
But it works exceedingly poorly; The ultimate solution is probably to modify the JavaScript, get the margins, padding, and exact position of the element while it's part of the document flow, and then explicitly set those values on the fixed position container.

A pure CSS solution would've been nice, though.

**Comment by jboxman** [comment]
Change to be made here: https://github.com/tscanlin/tocbot/blob/31c0fac3ccc18e618bf33951e0ad0b4a7174e606/src/js/build-html.js#L133.

The downside is someone may have JavaScript disabled. It's possible to hide the TOC, but that probably mangles the page because the column widths percentages assume that column is displayed.

But maybe we don't care about non-JavaScript users?

So that's the casualty here.

**Comment by IronicBadger** [comment]
This is awesome!!!!!!!

**Comment by vikram-redhat** [comment]
Love it!! Well done @jboxman!

As you mention the floating TOC needs to link back to the sections since that is such an expected behaviour that I worry we will get issues filed against that immediately :). I am not sure how much work that is though as you are then likely to need to interface between asciibinder output as a preprocessor action or post processing and a completely javascript hack. Let me know if you want to brainstorm on either.

**Comment by jboxman** [comment]
I can confirm that the JavaScript specific solution works:

```
var x = document.getElementById('js-toc-container');
undefined
var rect = x.getBoundingClientRect();
undefined
x;
<div id=​"js-toc-container">​…​</div>​
rect;
DOMRect {x: 825.328125, y: 175, width: 161.65625, height: 975, top: 175, …}
x.getBoundingClientRect();
DOMRect {x: 960, y: 0, width: 167.65625, height: 975, top: 0, …}
x.setAttribute('style', "position:fixed; left:825.328px;top:0;");
undefined
```

@vikram-redhat,

The larger issue is, regardless of whether the CSS-only solution worked (it doesn't really), the use of JavaScript here means that if someone has JavaScript disabled, we have an empty third column and the content will thus be compressed. I'm not sure how to deal with that, exactly, except use JavaScript to inject the third column dynamically, so it looks proper without JavaScript. This is kind of nasty, but certainly possible.

Meanwhile, there might be two options for the necessary change to the original script:

1. Modify the source and host it locally
1. Watch for DOM mutations with the [MutationObserver API](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) and dynamically set and remove the attribute with a local script (This only works for IE11+)

**Comment by jboxman** [comment]
Sometimes I have a dangerous obsession with solving a problem. I'm going to make breakfast now.

**Comment by vikram-redhat** [comment]
Glad the breakfast was nice :D 

While I am conscious of making it usable for folks who disable JavaScript, I want to make sure that the effort required to make that happen doesn't overshadow what we are trying to achieve here (and hereby also increase the complexity of the solution).

If JavaScript is disabled by someone browsing docs.openshift.com, we have other problems to worry about, not just this. For example, the main TOC itself wouldn't work :) 

Soooo.. let's not worry about that (and perhaps we have a minimum requirement posted somewhere for folks who do try and visit with Javascript disabled). 

The mutation solution is very interesting and I am not exactly sure how you are mapping that to the links, but I saw it works in the codepen link and that's a big thumbs up from me. Whenever I have used mutations in the past, I have ended up in horrible redirect loops, that says more about my programming skills though than the quality of this solution.

Anyhow, are you worried about the viewport resizing _after_ the user has loaded the page? Or are you trying to cater for all possible viewports?

**Comment by jboxman** [comment]
@vikram-redhat, it probably makes more sense if you examine the code in the _CodePen_.

Anyway, the resizing also happens when you've got a debugger pane open. So I opened it and the element suddenly was off screen, and I realized the entire document was repainted because the window was smaller. So this deals with the (edge) case where someone resizes the window; I'm sure it happens occasionally.

If you resize the pen now, the live TOC will move with the rest of the content as you'd expect if it were part of the normal DOM.

As for the mutation, that's just a cheat so I can see when js-toc JavaScript adds or removes the CSS class that fixes its position; When that happens, we need to set a few extra things directly on the element so it maintains the same position relative to the left side of the browser window. That's really what this entire quest is about: What's the `left` value for the element? With the legacy CSS-only solution I tried to reliably guess, but using a `margin-left` isn't viable. (It overlaps the page content, making it non-interactive.) The JavaScript solution merely grabs the element pixel position and directly injects it.

At this point I'm reasonably confident this solution works -- others ought to poke it with a stick, because I might have nonetheless overstated my success here.

(So hilariously there's a CSS `position: sticky` which does exactly what all this does, but it's behind a feature flag in many Firefox versions, uses a prefix on Safari, and isn't supported on any version of IE.)

I hadn't even realized the doc-set TOC used JavaScript. We can actually re-do that so it uses CSS only, for what it's worth.

**Comment by vikram-redhat** [comment]
Cool cool - I am nervous about using a fixed element pixel position that we may not always be able to control. The docs.openshift.com CSS does work with other CSS (although we override it in many instances). 

But if you think that that is a viable and reliable fix, let's go with it. 

Any chance we can demo to the team tomorrow? Along with @lbarbeevargas's https://github.com/openshift/openshift-docs/pull/24345 collapsible code segments, this would be a neat feature to launch 4.6 with.

**Comment by jboxman** [comment]
@vikram-redhat, PTO Monday.

**Comment by vikram-redhat** [comment]
> @vikram-redhat, PTO Monday.

Ah ok - I will drop in the link to the demo in Slack and the meeting so we can have a quick discussion to see if this is something we definitely want to go forward with. 

Enjoy your PTO!

**Comment by bergerhoffer** [comment]
@jboxman This is awesome!!! 

Just a few pieces of feedback after my first glance:

* I'm not a fan of the fact that adding this cuts down on the width of the main content area. With how narrow it is now, it reminds me too much of how the current customer portal docs are displayed, and I've always hated how narrow they are. How wide d.o.c docs are has always been a big reason for me to like them there over the portal docs. Is there any way to avoid this?
* Can we get some vertical padding between each section? Right now it's hard to tell where one section's text ends and the next one starts.
* I can't see the last few items at the end of the scrolling TOC. It ends for me with "Logging in to the cluster"

**Comment by lamek** [comment]
@jboxman this looks great. Suggestion to update the spacing to it's easier to read the titles and add a toggle so viewers can view or hide the "in-page toc" as wanted. Here is an example of a good toggle - https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/TroubleshootingInstances.html

(Bonus - this page also shows a list of in-page headers that runs off screen, and at that point the in-page toc becomes scrollable.)

**Comment by jboxman** [comment]
So as this progresses, it becomes ever more brittle. The white spacing and distinctiveness of each discrete TOC element can be addressed.

The content overflow, I'm not so sure as is. Because what's the correct behavior here? In practice, we have two distinct scenarios here to deal with:

- If the page is not scrolled past the mast height, the TOC container exists in the document flow
- If the page is scrolled beyond that, the TOC container is fixed position and exists outside the document flow (so it stays in one place while the page scrolls)

It's worse; On further reflection, what triggers an overflow is content exceeding the dimensions of the container. When is that true? Here it's variable. When a sub-tree expands, maybe it's too long and goes off page; but then maybe not for a smaller sub-tree.

Only display a scroll bar for content that _might_ be too long? Then we need to render a copy of the entirely expanded TOC, in a `display: hidden` element, measure its length, and see if it's long enough to warrant a permanently visible scroll bar. That requires code and might be brittle.

Correct handling for that isn't part of the js-toc JavaScript that actually draws the TOC, so it isn't a challenge the author sought to solve, or perhaps was unaware of.

Maybe the height needs to be the height of the browser window?

The path here is not entirely clear.

**Comment by jboxman** [comment]
Wrap attempt, smaller column sizes. Not ideal.

Screen Shot 2020-10-24 at 2 03 43 PM

**Comment by jboxman** [comment]
And our layout wasn't designed to accommodate this as cleanly as we'd prefer. The footer is absolutely placed on the bottom, and the TOC clips beneath it.
Screen Shot 2020-10-24 at 2 12 51 PM

**Comment by jboxman** [comment]
@openshift/team-documentation 

So we're also getting into larger issues about content presentation itself, which I wanted to avoid. For example, Amazon overflows their documentation, not their TOC. (So you scroll their technical content within a pane, rather than scrolling the entire page itself; This does make their TOC implementation easy; it scrolls in a pane too.) Our page length is instead determined by the length of our content, like a traditional blog post or news article on the Web.

(Actually, what they did is radically different from our layout -- The in-page TOC scroll bar is the _browser_ scroll bar; The others are content overflow scroll bars. That's a *vastly* different layout than we have, and not possible with our OCP docs layout today. As you scroll the content, the in-page TOC active header is eventually offscreen; They're okay with that trade-off.)

I wouldn't want to make substantial changes to any of this without measured deliberation. Over the past 5 or whatever odd years, users have come to expect a certain layout for OCP docs, and giving them some other layout will be jarring, to say the least.

Short of that, any in-page TOC is going to have some limitations that, if added to OCP docs, we'll have to live with. So is it worth the trade-off? There are certainly mitigations possible:

- If there are too many topics such that the in-page TOC overflows, we can restrict that page to h2s only, no h3s.
- We can hide the in-page TOC for smaller screen sizes, in addition to mobile; In practice this probably means if you aren't at least on a 15" display, it's hidden. (Minimum browser with probably 1200 pixels+.)

The more edge cases we try to address, the more likely we end up with a solution that's brittle. Unless we want to adopt a new documentation layout (CSS), however, this is the sandbox we have to play in.

Screen Shot 2020-10-24 at 2 32 46 PM

**Comment by jboxman** [comment]
In any case, I updated the CodePen to adjust some of the spacing concerns for the in-page TOC list itself.

**Comment by vikram-redhat** [comment]
Hey @jboxman - 

> We can hide the in-page TOC for smaller screen sizes, in addition to mobile; In practice this probably means if you aren't at least on a 15" display, it's hidden. (Minimum browser with probably 1200 pixels+.)

Might be good to see some data on this to help make decisions for screen resolutions. 

Screen Shot 2020-10-25 at 7 58 56 am

Screen Shot 2020-10-25 at 8 05 18 am

Based on that, I am inclined to say that folks reading the docs on mobile and smaller resolutions is a tiny fraction and therefore, ok to hide the TOC there.

> We can add a toggle to hide/reveal the in-page TOC. With some JavaScript work this is barely possible, but somewhat complicated. (Our column widths must be reset as hiding the in-page TOC brings us back to 2 columns, dynamically, from 3 columns. Our layout wasn't designed for this, so JavaScript is necessary to bridge the gap here. And it must take into account whether the in-page TOC is currently bound or fixed position.)

Yeah this could be jarring if not done cleanly. Do we have to resize the columns on hide and reveal?

**Comment by jboxman** [comment]
@vikram-redhat, we're using a legacy 2 column float layout. This adds a third column for the in-page TOC, which is necessary for the columns to align properly. The width of each is a % of the total area, and that area is defined based on a range of maximum viewable area. (For mobile, small, medium, and large sizes, where the size is a hard coded pixel value in our existing CSS.)

Given columns A, B, C, with 25%, 50%, and 25% width, given I can simply `display:none` column C, column A and B are still going to take up 75% of the space, inclusive of the now not displayed column. To fix this, JavaScript needs to re-set the classes on A and B that define the width. (And maybe set the width to `0` on C.) So instead of `col-md-3` and `col-md-7`, these need to be `col-md-4` and `col-md-8` respectively. (Our CSS layout uses a *grid* of up to 12 increments.)

So that must be done (or undone) whenever someone toggles the in-page TOC for display. For smaller screen sizes, this needs to happen as well, because the HTML as written includes all 3 columns by default in this implementation. The alternate is to keep the existing layout, and *inject* the additional column and reset the classes on page render.

So there are no ideal answers for this. The disposition of the columns at the beginning is merely the initial starting point; The same work needs be done regardless when hiding or displaying the in-page TOC. It's easier to avoid injecting HTML whenever possible, and run with the document as it is.

(The CSS media query seems to do this automatically if you shrink the window small enough to be 'mobile', but I'm not sure if I can trigger that in some other way.)

**Comment by vikram-redhat** [comment]
> Given columns A, B, C, with 25%, 50%, and 25% width, given I can simply `display:none` column C, column A and B are still going to take up 75% of the space, inclusive of the now not displayed column. To fix this, JavaScript needs to re-set the classes on A and B that define the width. (And maybe set the width to `0` on C.) So instead of `col-md-3` and `col-md-7`, these need to be `col-md-4` and `col-md-8` respectively. (Our CSS layout uses a _grid_ of up to 12 increments.)

You are assuming that we _need to_ redistribute the width. What if we were to not redistribute the width but simply do nothing on hiding of col c? 

The reason to hide col c (from a reader's perspective) is two folds: a) make the rest of the sections wider for easy reading b) reduce the amount of clutter on screen for less distracted reading. 

> So that must be done (or undone) whenever someone toggles the in-page TOC for display. For smaller screen sizes, this needs to happen as well, because the HTML as written includes all 3 columns by default in this implementation. The alternate is to keep the existing layout, and _inject_ the additional column and reset the classes on page render.

Why do we need to inject HTML in all 3 columns? We can detect the screen size at page load time, and redistribute or not. Check out the Ruby templates for the pages.

*sigh* But other than that, I think the better solution is to redesign our CSS and design, which is a bigger project. I don't know, at this stage, which way I prefer.

**Comment by jboxman** [comment]
@vikram-redhat, redistribution goes to Andrea's point, which is how crowded the page is with this addition. Offering the option to hide it at larger resolutions ought to recover the space it uses.

The issue of HTML inject or not pertains specifically to the issue of whether the Ruby template which generates the HTML output we deliver to the user is a 2 column or 3 column layout. If the former, in order to have an in-page TOC, it must go somewhere, and this design relies on it living in a 3rd column. So it must be added. If it already exists, as it does in the CodePen demo, nothing special needs to be done.

But in either case, if we wanted to allow someone to toggle the visibility of the in-page TOC for the purposes of increasing the available space for documentation content, we need the browser to adjust the column widths when the TOC is hidden or revealed. (Of course if we don't care, the in-page TOC can always be visible and none of this actually matters. The layout automatically switches to a single column layout for mobile, the site TOC is available with a toggle, and the in-page TOC is simply hidden without anyone being the wiser.)

But we're probably talking past each other on this; We'll have to review it on Monday.
